### PR TITLE
Add JSON field type and Event object

### PR DESCRIPTION
**Pull request by FelixMalfait**
Stopping the PR here because it's functional and already big
It covers:
- Add a new field type: JSON
- Rename CreateEvent (frontend) to Track
- Add a new standard object: Event
- Create listener/object to log create/update/delete on companies/people/opportunities 

But we still need to:
- add featureFlag (needed before merge, the rest isn't)
- frontend to display logs on show page
- support events on custom objects and maybe more standard objects (e.g. webhooks)
- log user who performed the action
- check if event are properly sent for updateMany
- refactor webhooks to be sent through events not triggerWebhooks
- frontend for the json field type (allow creation of json field, and display json nicely in the ui)

https://github.com/twentyhq/twenty/assets/6399865/af535ba8-71eb-499b-b715-9e2de78e4550

### CODE REVIEW

File: packages/twenty-front/src/generated/graphql.tsx [hunk 1]
```diff
@@ -492,6 +493,23 @@ export type RelationConnection = {
   pageInfo: PageInfo;
 };
 
+export type RelationDefinition = {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/generated/graphql.tsx:496]: This is not related to this PR

File: packages/twenty-front/src/generated-metadata/graphql.ts [hunk 2]
```diff
@@ -66,10 +66,32 @@ export type AuthTokens = {
 export type Billing = {
   __typename?: 'Billing';
   billingFreeTrialDurationInDays?: Maybe<Scalars['Float']['output']>;
-  billingUrl: Scalars['String']['output'];
+  billingUrl?: Maybe<Scalars['String']['output']>;
   isBillingEnabled: Scalars['Boolean']['output'];
 };
 
+export type BillingSubscription = {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/generated-metadata/graphql.ts:73]: This is not related to this PR - came in the graphql generation

File: packages/twenty-front/src/generated-metadata/graphql.ts [hunk 3]
```diff
@@ -631,6 +654,23 @@ export type RelationConnection = {
   pageInfo: PageInfo;
 };
 
+export type RelationDefinition = {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/generated-metadata/graphql.ts:657]: This is not related to this PR - came in the graphql generation

File: packages/twenty-front/src/generated-metadata/graphql.ts [hunk 4]
```diff
@@ -347,12 +370,6 @@ export type MutationCheckoutSessionArgs = {
 };
 
 
-export type MutationCreateEventArgs = {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/generated-metadata/graphql.ts:350]: Had to rename "createEvent" which was a function used mainly for telemetry (might one day be used for user-facing audit logs too). The name was conflicting with the new event object

File: packages/twenty-server/src/engine-metadata/field-metadata/dtos/default-value.input.ts [hunk 5]
```diff
@@ -16,6 +17,12 @@ export class FieldMetadataDefaultValueString {
   value: string | null;
 }
 
+export class FieldMetadataDefaultValueJson {
+  @ValidateIf((_object, value) => value !== null)
+  @IsJSON()
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine-metadata/field-metadata/dtos/default-value.input.ts:22]: I haven't tested extensively everything related to the JSON field type. It will be easier to test once we build the frontend in a followup ticket

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/record-position.listener.ts [hunk 6]
```diff
@@ -21,27 +20,31 @@ export class RecordPositionListener {
 
   @OnEvent('*.created')
   async handleAllCreate(payload: ObjectRecordCreateEvent<any>) {
-    if (!hasPositionField(payload.createdObjectMetadata)) {
+    if (!hasPositionField(payload.objectMetadata)) {
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/record-position.listener.ts:23]: See below, the reason there are a lot of renaming is that there was not standard convention for events related to entities creation/update/delete, it was getting messy as we were not always passing the same prop depending on context. I tried to improve this a bit. I feel like it's better, but could still be improved in a future iteration

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts [hunk 7]
```diff
@@ -328,6 +334,8 @@ export class WorkspaceQueryRunnerService {
       options,
     );
 
+    // TODO: check - NO EVENT SENT?
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:337]: Unless I'm missing no event was sent for updateMany? I guess we just forgot?

File: packages/twenty-server/src/engine/api/graphql/workspace-schema-builder/graphql-types/input/json-filter.input-type.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { GraphQLInputObjectType } from 'graphql';
+
+import { FilterIs } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/input/filter-is.input-type';
+
+export const JsonFilterType = new GraphQLInputObjectType({
+  name: 'JsonFilter',
+  fields: {
+    is: { type: FilterIs },
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/api/graphql/workspace-schema-builder/graphql-types/input/json-filter.input-type.ts:8]: Not sure if this is relevant, this is the type of thing I was referring to when I said it will be easier to test when we have the frontend

File: packages/twenty-server/src/engine/integrations/message-queue/message-queue.constants.ts [hunk 9]
```diff
@@ -9,4 +9,5 @@ export enum MessageQueue {
   calendarQueue = 'calendar-queue',
   billingQueue = 'billing-queue',
   recordPositionBackfillQueue = 'record-position-backfill-queue',
+  entityEventsToDbQueue = 'entity-events-to-db-queue',
```
Comment (Weiko) [inline packages/twenty-server/src/engine/integrations/message-queue/message-queue.constants.ts:12]: Same comment as above. I think we should clearly define that those are events from the API and not DB so later when we will use real DB events it won't be confusing

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/jobs/save-event-to-db.job.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    if (
+      data.objectName != 'person' &&
+      data.objectName != 'company' &&
+      data.objectName != 'opportunity'
+    ) {
+      return;
+    }
+
+    await workspaceDataSource?.query(
+      `INSERT INTO ${dataSourceMetadata.schema}."event"
```
Comment (Weiko) [inline packages/twenty-server/src/engine/api/graphql/workspace-query-runner/jobs/save-event-to-db.job.ts:47]: I think we want to save the workspaceMember too?

Comment (Weiko) [reply packages/twenty-server/src/engine/api/graphql/workspace-query-runner/jobs/save-event-to-db.job.ts:47]: Ok my bad I saw the TODO above

**Review by Weiko — COMMENTED** [review]
Left some comments @FelixMalfait

File: packages/twenty-front/src/modules/analytics/graphql/queries/track.ts [hunk 11]
```diff
@@ -1,8 +1,8 @@
 import { gql } from '@apollo/client';
 
 export const CREATE_EVENT = gql`
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-front/src/modules/analytics/graphql/queries/track.ts:3]: Thanks!

File: packages/twenty-front/src/testing/graphqlMocks.ts [hunk 12]
```diff
@@ -34,7 +34,7 @@ export const graphqlMocks = {
     graphql.mutation(getOperationName(CREATE_EVENT) ?? '', () => {
       return HttpResponse.json({
         data: {
-          createEvent: { success: 1, __typename: 'Event' },
+          track: { success: 1, __typename: 'Event' },
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-front/src/testing/graphqlMocks.ts:37]: Thanks!

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+import { MessageQueue } from 'src/engine/integrations/message-queue/message-queue.constants';
+import { MessageQueueService } from 'src/engine/integrations/message-queue/services/message-queue.service';
+import { ObjectRecordCreateEvent } from 'src/engine/integrations/event-emitter/types/object-record-create.event';
+import {
+  SaveEventToDbJobData,
+  SaveEventToDbJob,
+} from 'src/engine/api/graphql/workspace-query-runner/jobs/save-event-to-db.job';
+
+@Injectable()
+export class EntityEventsToDbListener {
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts:13]: I wasn't too keen on this name either yes! But didn't find anything better.
For me it's pretty clear that this is not about listening DB events with _to_ (role is to send entity events *to* DB).
Any suggestion?

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts [hunk 14]
```diff
@@ -242,10 +242,11 @@ export class WorkspaceQueryRunnerService {
     parsedResults.forEach((record) => {
       this.eventEmitter.emit(`${objectMetadataItem.nameSingular}.created`, {
         workspaceId,
-        createdRecord: this.removeNestedProperties(record),
-        createdObjectMetadata: {
-          nameSingular: objectMetadataItem.nameSingular,
-          isCustom: objectMetadataItem.isCustom,
+        recordId: record.id,
+        objectMetadata: objectMetadataItem,
+        operation: 'created',
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:247]: yes good point!

File: packages/twenty-server/src/engine/integrations/message-queue/message-queue.constants.ts (hunk 9 above)
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/integrations/message-queue/message-queue.constants.ts:12]: (same discussion / it says events TO db but definitely open to better naming!)

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  constructor(
+    @Inject(MessageQueue.entityEventsToDbQueue)
+    private readonly messageQueueService: MessageQueueService,
+  ) {}
+
+  @OnEvent('*.*')
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts:19]: Good point! 
Replacing by  
@OnEvent('*.created')
 @OnEvent('*.updated')
 (realized thanks to your comment that deleted wasn't a good idea as long as we don't have soft-delete)

File: packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Weiko) [reply packages/twenty-server/src/engine/api/graphql/workspace-query-runner/listeners/entity-events-to-db.listener.ts:13]: Aaah, I missed the to. So basically it's for events of queries that go to the DB. Mmm, I would have called it WorkspaceQueryRunnerEventsListener I think 😅 @FelixMalfait but yours is fine too, I didn't see the "to" the first time
